cmd/snap-confine/mount-support.c

```diff
@@ -578,6 +578,18 @@ void sc_populate_mount_ns(struct sc_apparmor *apparmor, int snap_update_ns_fd,
 					die("cannot locate the core or legacy core snap (current symlink missing?)");
 				}
 			}
+			if (sc_streq(base_snap_name, "core16")) {
```

Comment:
As a new comment, by the time this is determined and base is changed there has already been some code that ran in `snap-confine.c` that used the original value (core16).

I'm very concerned with this change. I would like to understand why we cannot make this change earlier, in snap-run, like I described above. Regarding the compatibility code for ubuntu-core, that code is almost never used now. and might be just buggy.

I would love to consider removing that logic from snap-confine entirely.

Reply:
As I explained we cannot do the change naively in snap run, because we do want to pivot, but use core instead of core16. We can do a change in snap run, but something would still need to change in snap-confine.

About your comment, you are correct that  sc_join_preserved_ns  as well (which is the second path that ends up calling   sc_should_use_normal_mode)  should behave consistently with this code. And right now it wouldn't.  So we need more checks/cases in the spread test.

Reply:
So I'm sure the code has grown organically, but we should not be making decisions about the fundamental mode of operation down levels and repeatedly in the code. It seems we need a refactor where we make the mode and actual base choice in snap-confine.c itself, and then instead of carrying the base_snap_name, we carry a small option struct with useful information about the mode and base.